src/nanotron/models/llama.py

```diff
@@ -764,6 +776,19 @@ def forward_with_hidden_states(
         for encoder_block in self.decoder:
             hidden_encoder_states = encoder_block(**hidden_encoder_states)
 
+        # hidden_states.shape = [seq_length/tp_rank, batch_size, hidden_dim]
+        mup_l1_norm = hidden_encoder_states["hidden_states"].mean(dim=[0, 1]).abs()  # [hidden_dim]
+        dist.all_reduce(
+            mup_l1_norm, op=dist.ReduceOp.SUM, group=self.parallel_context.tp_pg
+        )  # sum [hidden_dim] across tp ranks
+        mup_l1_norm = mup_l1_norm.mean()
+        dist.all_reduce(mup_l1_norm, op=dist.ReduceOp.AVG, group=self.parallel_context.dp_pg)
+
+        if dist.get_rank() == 0:
+            import wandb
+
+            wandb.log({"output_l1_norm": mup_l1_norm.cpu().detach().float().numpy(), "width": self.config.hidden_size})
```

Comment:
this adds a cpu-gpu sync. can we return the values we want to log as outputs for now, and log them from trainer (after train_batch_iter)

Reply:
later on it would be great to have an attachable Mixin class where we can add tensors that we want to log after the end of train_batch_iter